### DISCUSSION TITLE
erf(x) support

### DISCUSSION THREAD

**Discussion by kkaryos** [vgvassilev/clad#869] [Q&A] [answered]
Can we use clad to differentiate erf(x)?

**Comment by alexander-penev** [comment, marked as answer]
Yes.

It is not very well documented yet, but when there is a built-in function (or other) that is not yet supported by clad, we can use the so-called custom_derivatives. This is done as follows:

``` C++
#include "clad/Differentiator/Differentiator.h"

#include <iostream>
#include <math.h>

namespace clad { namespace custom_derivatives {
namespace std {
  template <typename T>
  ValueAndPushforward<T, T> erf_pushforward(T x, T d_x) {
    return {::std::erf(x), M_2_SQRTPI*::std::exp(-::std::pow(x,2.0)) * d_x};
  }
}
using std::erf_pushforward; // Required if we use C style function call erf(...) instead C++ std::erf(...)
}}

double f(double x) {
  return std::erf(x) + 1;
}

int main() {
  auto df = clad::differentiate(f, 0);
  df.dump();
  std::cout << "f(3)=" << f(3.0) << "; df(3)=" << df.execute(3.0) << std::endl;

  return 0;
}
```
If we compile it with

```clang++ -Xclang -add-plugin -Xclang clad -Xclang -load -Xclang clad.so -I../include/ -std=c++11 erf.cpp```

and run the program we will get a working program in which `erf(...)` is differentiated with the custom defined function `erf_pushforward(...)`:

```
The code is:
double f_darg0(double x) {
    double _d_x = 1;
    ValueAndPushforward<double, double> _t0 = clad::custom_derivatives::erf_pushforward(x, _d_x);
    return _t0.pushforward + 0;
}

f(3)=1.99998; df(3)=0.000139253
```

If this function is not defined, clad generates a warning and uses numeric differentiation for `erf`.

```
warning: Falling back to numerical differentiation for 'erf' since no suitable overload was found and clad could not derive it. To disable this feature, compile your programs with -DCLAD_NO_NUM_DIFF.

The code is:
double f_darg0(double x) {
    double _d_x = 1;
    return numerical_diff::forward_central_difference(std::erf, x, 0, 0, x) * _d_x + 0;
}

f(3)=1.99998; df(3)=0.000139253
```

Of course, in the near future we plan to implement custom derivatives for all built-in functions, as part of a clad, as has already been done for some of the functions in math.h [here](https://github.com/vgvassilev/clad/blob/d879f1b8092714b209790b17aa1ed1704d0eabc9/include/clad/Differentiator/BuiltinDerivatives.h#L62)

**Reply by kkaryos** [reply]
Thank you very much for your response, very helpful!
 I also wanted to ask about if statements and for loops. It is stated in the documentation that they are not yet supported. Is there an estimated timeframe for rolling out support?

**Reply by alexander-penev** [reply]
Rather, the documentation is not very up-to-date. If statements and loops have been supported before version 0.1 of clad. Of course, it is possible that in some very specific cases there will be problems and bugs (it depends on which differential mode you will use - for example, in forward mode almost everything should work).
To be sure if the AD function is correct, it is good to check its results in another way. I often use numerical differentiation, despite its instability, to confirm the results - if they match everything is OK, if not - a deeper check should be done :)

**Reply by alexander-penev** [reply]
By the way, you can also get an idea of what is implemented and what works for sure by looking at the examples and tests in the clad source itself.

**Reply by vgvassilev** [reply]
We will need the input code that caused the crash.
